CONTRIBUTING.md

```diff
@@ -1,27 +1,92 @@
+Thank you for investing your time in contributing to our project!
 
-Before starting, we encourage you to take a look to the following documentation as it might help you in your first contribution adventure:
-- https://github.com/privacy-scaling-explorations/zkevm-docs
-- https://github.com/privacy-scaling-explorations/zkevm-specs
+# How you can Contribute
+
+First take a look at our [specs](https://github.com/privacy-scaling-explorations/zkevm-specs), [auditing videos](https://www.youtube.com/watch?v=HhHTho2QZa4), and [zkevm docs](https://github.com/privacy-scaling-explorations/zkevm-docs). It’s very important that you understand at a high-level how the ZK-EVM architecture and all of its components work.
+
+## Best Coding Practices
+
+### Creating PRs
+
+1. Fork the zkevm or specs repos.
+2. Write a simple markdown spec doc about the circuit PR and send to the `zkevm-specs` repo.
+3. For PRs that change significantly the codebase or the way on which some procedure is done (test-suite, cross-crate structures, module refactors) please [file an issue](https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/new/choose) and wait until this feature/proposal gets approved by the repo maintainers.
+4. Send the circuit PR to the `zkevm-circuits` repo. Make sure you give a clear title and succinct description of the PR and the area you are changing. 
+    - ******************NOTE:****************** Make atomic PRs. Focused on a single feature or topic. Don't mix different topics or issues in the same PR if it can be avoided.
+5. Create a separate branch that will host all commits to the PR. Give branch names succinct title describing the change.
+6. Start to make your changes. Please also make sure that the following commands pass if you have changed the code:
+    
+    ```
+    make test
+    make clippy
+    ```
+    
+7. Once you made the changes, give the commit message a sufficient description of the change. The message content can be brief, including only enough details to summarize the purpose of the commit. In other instances, more in-depth explanations about how the commit accomplishes its goals are better suited as comments in the code.
+    - ******************NOTE:****************** The commits in a pull request should be organized in a manner where each commit represents a small, coherent step towards the pull request's overall objective. The structure of the pull request should facilitate the reviewer in easily tracking and understanding each change made.
+8. You may run into unnecessary commits, such as merge conflicts or minor refactorings. In that case, always `git rebase -i`  and [squash](https://www.git-tower.com/learn/git/faq/git-squash) all your commits to one commit. 
+9. Please do not merge master into your branch as you develop your pull request; instead, rebase your branch on top of the latest master if your pull request branch is long-lived.
+    1. How to regularly update your PR with master:
```

Comment:
If we assume the contributor needs this detailed git command guide, then squashing commits might be difficult for them. We can make the commits squashing optional.

Reply:
We must also consider that currently for outdated branches in a PR, github shows a button called "Update branch" that merges main into the branch (instead of doing the rebase).  This seems to be very convenient for reviewers who want to merge a PR from a contributor without write access; because after an approval they can just click, wait for tests, and merge.  Maybe we could rephrase points 8 and 9 as suggestions?